### DISCUSSION TITLE
[Feature Proposal] Support for arbitrary selector (variant)

### DISCUSSION THREAD

**Discussion by neupauer** [tailwindlabs/tailwindcss#6630] [Ideas]
# Support for arbitrary selector (variant)

I would like to suggest a feature to support an arbitrary selector.

It can come in handy in cases where we don't have control over the HTML.

I'll use an example to show what I mean by "arbitrary selector":

```html
<ul class="[li]:py-2">
  <li>01</li>
  <li>02</li>
  <li>03</li>
</ul>
```

```html
<article class="prose [table_a]:text-blue-500">
  {{ markdown }}
</article>
```

```html
<ul class="[li>a]:text-blue-500">
  <li>Link <a href="#">01</a></li>
  <li>Link <a href="#">02</a></li>
  <li>Link <a href="#">03</a></li>
</ul>
```

With other variants:

```html
<ul class="[li>a]:md:hover:text-blue-500">
  <li>Link <a href="#">01</a></li>
  <li>Link <a href="#">02</a></li>
  <li>Link <a href="#">03</a></li>
</ul>
```

The override should work:

```html
<ul class="[li]:text-red-500">
  <li>Link <a href="#">01</a></li>
  <li>Link <a href="#" class="text-blue-500">02</a></li>
  <li>Link <a href="#">03</a></li>
</ul>
```

**Comment by adamwathan** [comment]
Definitely on our radar! The stuff we added to the typography plugin recently like `prose-h1:underline` is a baby step towards working on this.

**Reply by neupauer** [reply]
wow, that's gonna be amazing

**Comment by joe-bell** [comment]
I was just about to raise a discussion post for the exact same thing! No urgency for this at all, but it would be wonderful

**Comment by sukima** [comment]
I've been a non-tailwind fan for some time and this missing feature is the main reason I've struggled to accept tailwind. When it lands I think I might finally endorse TW.

Is there a workaround we can use till this gets any headway?

**Comment by sntran** [comment]
I am also thinking of writing a plugin for `:has` pseudo selector, with signature similar to this:

```html
<a class="has-(>img):block"></a>
```

I was wondering why my function for `addVariant` was not called, and it looks like Tailwind does not consider `[>img]` as arbitrary value.

If arbitrary selector is supported, then it will open a lot of possibility.

**Comment by neupauer** [comment]
https://github.com/tailwindlabs/tailwindcss/pull/8299 🎉
